zerver/views/registration.py

```diff
@@ -861,6 +861,12 @@ def realm_redirect(request: HttpRequest, next: str = REQ(default="")) -> HttpRes
         if form.is_valid():
             subdomain = form.cleaned_data["subdomain"]
             realm = get_realm(subdomain)
+            # If a user visits just /accounts/go/, and is not logged in,
+            # the user should be redirected to login_page.
+            # Special handling is required because user gets directed to
+            # `web_public_view` in this case by default.
+            if next == "" and not request.user.is_authenticated:
+                next = settings.HOME_NOT_LOGGED_IN
```

Comment:
Is it correct to only do this when `next` was unset? I feel like if you clicked a link to a particular part of the Zulip app via this flow, and aren't logged in, you likely still want to be logged in before being taken to that page.

So it's possible we want to construct a new `next` that combines `HOME_NOT_LOGGED_IN` with the `next` from the existing page.

Reply:
This flow will always come into picture in the following scenarios:
visiting `/accounts/go/` or something of the kind `accounts/go/?next=something`

Everything for the case when we have `accounts/go/?next=something` is handled properly by the view corresponding to 
`subdomain.zulipchat.com/something`. i.e if the user is not logged in, enforce login.

For the case `accounts/go/`, the flow goes to the view corresponding to `subdomain.zulipchat.com` i.e `home` view.
So for the case when `next` is unset we need to enforce `login` as discussed in [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/log.20in.20flow.20w.2F.20web-public/near/1454922).

**Can you please give an example url for this (I'm still not clear with this line):**
> So it's possible we want to construct a new next that combines HOME_NOT_LOGGED_IN with the next from the existing page.

Reply:
Another possible way:
```python
@has_request_variables
def realm_redirect(request: HttpRequest, next: str = REQ(default="")) -> HttpResponse:
    if request.method == "POST":
        form = RealmRedirectForm(request.POST)
        if form.is_valid():
            subdomain = form.cleaned_data["subdomain"]
            realm = get_realm(subdomain)
            ```
            redirect_url = realm.uri
            if not request.user.is_authenticated:
                redirect_url = realm.uri + settings.HOME_NOT_LOGGED_IN
            redirect_to = get_safe_redirect_to(next, redirect_url)
            return HttpResponseRedirect(redirect_to)
            ```
    else:
        form = RealmRedirectForm()
       
    return render(request, "zerver/realm_redirect.html", context={"form": form})
  ```
  I think you were talkinng about this approach here:
  > So it's possible we want to construct a new next that combines HOME_NOT_LOGGED_IN with the next from the existing page
  
  **combining HOME_NOT_LOGGED_IN with the next from the existing page**
  
  Let me know if i'm right @timabbott

Reply:
Another possible way:
```python
@has_request_variables
def realm_redirect(request: HttpRequest, next: str = REQ(default="")) -> HttpResponse:
    if request.method == "POST":
        form = RealmRedirectForm(request.POST)
        if form.is_valid():
            subdomain = form.cleaned_data["subdomain"]
            realm = get_realm(subdomain)
            ```
            redirect_to = get_safe_redirect_to(next, realm.uri)
            if not request.user.is_authenticated:
                redirect_to = append_url_query_string(
                    realm.uri + settings.HOME_NOT_LOGGED_IN, urlencode({REDIRECT_FIELD_NAME: next})
                )
            return HttpResponseRedirect(redirect_to)
            ```
    else:
        form = RealmRedirectForm()

    return render(request, "zerver/realm_redirect.html", context={"form": form})
  ```
  I think you were talkinng about this approach here:
  > So it's possible we want to construct a new next that combines HOME_NOT_LOGGED_IN with the next from the existing page
  
  **combining HOME_NOT_LOGGED_IN with the next from the existing page**
  
  Let me know if i'm right @timabbott